zerver/lib/message.py

```diff
@@ -344,3 +347,86 @@ def render_markdown(message, content, realm=None, realm_alert_words=None, messag
         message.is_me_message = Message.is_status_message(content, rendered_content)
 
     return rendered_content
+
+def add_missing_messages(user_profile):
+    # type: (UserProfile) -> None
+    # This list will store all the messages for which we eventually will create
+    # UserMessage table rows by doing a bulk insert.
+    all_messages_to_insert = []  # type: List[UserMessage]
+
+    def store_user_message_to_insert(message):
+        # type: (Message) -> None
+        message = UserMessage(user_profile=user_profile,
+                              message_id=message['id'], flags=0)
+        all_messages_to_insert.append(message)
+
+    def find_and_store_to_insert_stream_msgs(all_stream_messages,
+                                             all_stream_subscription_logs):
+        # type: (DefaultDict[int, List[Message]], DefaultDict[int, List[RealmAuditLog]]) -> None
+        for (stream_id, stream_messages) in all_stream_messages.items():
+            stream_subscription_logs = all_stream_subscription_logs[stream_id]
+            for log_entry in stream_subscription_logs:
+                if len(stream_messages) > 0:
+                    if log_entry.event_type == 'subscription_deactivated':
+                        for stream_message in stream_messages:
+                            if stream_message['pub_date'] < log_entry.event_time:
+                                store_user_message_to_insert(stream_message)
+                            else:
+                                break
+                    else:
+                        initial_msg_count = len(stream_messages)
+                        for i, stream_message in enumerate(stream_messages):
+                            if stream_message['pub_date'] >= log_entry.event_time:
+                                stream_messages = stream_messages[i:]
+                                break
+                        final_msg_count = len(stream_messages)
+                        if initial_msg_count == final_msg_count:
+                            if stream_messages[-1]['pub_date'] < log_entry.event_time:
+                                stream_messages = []
+            if len(stream_messages) > 0:
+                for stream_message in stream_messages:
+                    store_user_message_to_insert(stream_message)
+
+    all_subs = Subscription.objects.select_related('recipient').filter(
+        user_profile=user_profile).values('recipient')
+
+    recipient_ids = [sub['recipient'] for sub in all_subs]
```

Comment:
I gave this a fair amount of thought and it feels like it might not be feasible to do this currently. Like it would be a total overkill. To find information about time when subscription was deactivated we would need to go to realmauditlog and then do comparison stuff based upon stream filtering and last_active_message_id. It seems pretty much repetitive. Instead I am proposing to rather not filter it here and achieve the optimisation effect desired here by later short circuiting by making use of a conditional in 'find_and_store_to_insert_stream_msgs'. I will tag the change which I did to address this when I push the update here.